core/src/main/scala/io/qbeast/spark/writer/RollupDataWriter.scala

```diff
@@ -138,21 +140,33 @@ trait RollupDataWriter extends DataWriter {
     rollup.compute()
   }
 
-  private def getRollupCubeIdUDF(
-      revision: Revision,
-      rollup: Map[CubeId, CubeId]): UserDefinedFunction = udf({ cubeIdBytes: Array[Byte] =>
-    val cubeId = revision.createCubeId(cubeIdBytes)
-    var rollupCubeId = rollup.get(cubeId)
-    var parentCubeId = cubeId.parent
-    while (rollupCubeId.isEmpty) {
-      parentCubeId match {
-        case Some(value) =>
-          rollupCubeId = rollup.get(value)
-          parentCubeId = value.parent
-        case None => rollupCubeId = Some(cubeId)
+  private def computeCubeUUIDs(rollup: Map[CubeId, CubeId]): Map[CubeId, String] = {
+    val uuidCache = mutable.Map[CubeId, String]()
+    rollup.foreach { case (_, cubeToRollupId) =>
+      if (!uuidCache.contains(cubeToRollupId)) {
+        uuidCache(cubeToRollupId) = UUID.randomUUID().toString
       }
     }
-    rollupCubeId.get.bytes
-  })
+    uuidCache.toMap
+  }
+
+  private def getFileUUIDUDF(
```

Comment:
Using two maps for a UDF can add significant overhead.
You can merge `rollup` and `rollupCubeUUIDs` : `rollup + rollupCubeUUIDs: Map[CubeId, String]`

Reply:
Fixed